### ISSUE TITLE
Consider using official use-subscription package 

### ISSUE THREAD

**Issue by abliarsar** [effector/effector#212]
Currently, useStore hook subscribes in render phase. That makes it not compatible with future concurrent mode changes.

## Proposal
Reimplement useStore with official useSubscription hook provided by react team.

## Use case
Avoid subscribing in render phase.
Make it fully compatible with future react changes. 

[use-subscription package](https://github.com/facebook/react/blob/master/packages/use-subscription)
I made a small [codesandbox](https://codesandbox.io/embed/great-babbage-g02pw) and it seems that edge case for dispatching in child components before parent subscription works fine.

Event if you don't want to rely on extra dependency, we can use similar techniques to make useStore compatible with concurrent mode.

@zerobias, what do you think?

**Comment by zerobias** [comment]
Interesting, `useSubscription` package implemented _exactly_ like `useStore`, even with [the same](https://github.com/facebook/react/blob/master/packages/use-subscription/src/useSubscription.js#L72) unsubscription [flag](https://github.com/zerobias/effector/blob/master/src/react/useStore.js#L26)

1. If we add `store.getCurrentValue = store.getState`, then we'll be able to use `useSubscription` [directly](http://codesandbox.io/s/frosty-lamport-e8n2f)
```js
const useStore = store => {
  store.getCurrentValue = store.getState
  return useSubscription(store)
}
``` 
2. The one significant difference is [isomorphicLayoutEffect](https://github.com/zerobias/effector/blob/master/src/react/useStore.js#L29) instead of [useEffect](https://github.com/facebook/react/blob/master/packages/use-subscription/src/useSubscription.js#L70). Not sure that I understand their arguments against sync unsubscription 🤔

Need to ensure that replacing the synchronous useLayoutEffect with useEffect will not cause unforeseen issues

**Comment by abliarsar** [comment]
Oops, accidently closed this one, sorry. 
useEffect should not cause any problems, i think, but it should be somehow tested, yeah (but how?). 
They are using passive listener because they want to subscribe asynchronously. So async unsubscribtion is just a side effect. We should not really think, when component will unsubscribe, because now sub/unsub is parts of one effect and react guarantees, that there is no memory leaks.

**Comment by zerobias** [comment]
Seems there is very subtle but measurable [difference](https://codesandbox.io/s/quirky-sky-9dn5y)

In case with very frequent updates, some of them able to "seep" between react phases, which leads to rerendering, if that example is correct

![image](https://user-images.githubusercontent.com/15912112/66717606-836f9d80-ede3-11e9-9253-ee28a6d69f2f.png)

**Comment by zerobias** [comment]
Seems like difference was fixed between react 16.8.6 and current 16.12.0. But despite this, the team already had [a new idea](https://github.com/bvaughn/rfcs/blob/useMutableSource/text/0000-use-mutable-source.md), so let them decide what they can offer for us at first.
